File: packages/contracts-python/src/aegis_contracts/approvals.py
"""Phase 24 approval workflow contracts."""

from __future__ import annotations

from enum import StrEnum
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, model_validator

from aegis_contracts.entities import ApprovalV1, ExecutedActionV1, ProposalStatus
from aegis_contracts.errors import ContractErrorCode, ContractValidationError
from aegis_contracts.primitives import (
    ApprovalId,
    EventId,
    ProposalId,
    Revision,
    RunId,
    UtcTimestamp,
)
from aegis_contracts.proposals import (
    PolicyDecisionV1,
    PolicyOutcomeV1,
    ResponseOptionV1,
    ScenarioCommandTemplateV1,
)
from aegis_contracts.versioning import (
    APPROVE_PROPOSAL_REQUEST_SCHEMA_VERSION,
    APPROVE_PROPOSAL_RESPONSE_SCHEMA_VERSION,
    AUTHORIZED_SIMULATION_COMMAND_SCHEMA_VERSION,
    CANCEL_PROPOSAL_REQUEST_SCHEMA_VERSION,
    EXECUTION_RESULT_SCHEMA_VERSION,
    FINAL_POLICY_CHECK_SCHEMA_VERSION,
    MODIFY_PROPOSAL_REQUEST_SCHEMA_VERSION,
    MODIFY_PROPOSAL_RESPONSE_SCHEMA_VERSION,
    PROPOSAL_MODIFICATION_SCHEMA_VERSION,
    REJECT_PROPOSAL_REQUEST_SCHEMA_VERSION,
    REJECT_PROPOSAL_RESPONSE_SCHEMA_VERSION,
    STALE_PROPOSAL_ERROR_SCHEMA_VERSION,
    assert_supported_schema_version,
)


class ApprovalErrorCode(StrEnum):
    STALE_PROPOSAL = "STALE_PROPOSAL"
    UNAUTHORIZED = "UNAUTHORIZED"
    POLICY_BLOCKED = "POLICY_BLOCKED"
    VALIDATION_FAILED = "VALIDATION_FAILED"
    CONFLICT = "CONFLICT"
    NOT_FOUND = "NOT_FOUND"
    ALREADY_DECIDED = "ALREADY_DECIDED"
    EXECUTION_FAILED = "EXECUTION_FAILED"


class StaleProposalErrorV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    code: ApprovalErrorCode = Field(default=ApprovalErrorCode.STALE_PROPOSAL)
    message: str = Field(min_length=1, max_length=1024)
    proposal_id: ProposalId = Field(alias="proposalId")
    expected_revision_id: str | None = Field(default=None, alias="expectedRevisionId")
    actual_revision_id: str | None = Field(default=None, alias="actualRevisionId")
    expected_revision: Revision | None = Field(default=None, alias="expectedRevision")
    actual_revision: Revision | None = Field(default=None, alias="actualRevision")
    details: dict[str, Any] = Field(default_factory=dict)

    @model_validator(mode="after")
    def validate_schema_version(self) -> StaleProposalErrorV1:
        assert_supported_schema_version("stale_proposal_error", self.schema_version)
        if self.schema_version != STALE_PROPOSAL_ERROR_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported stale proposal error schema: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self


class ProposalModificationV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    proposal_id: ProposalId = Field(alias="proposalId")
    previous_revision_id: str = Field(alias="previousRevisionId", min_length=1, max_length=64)
    new_revision_id: str = Field(alias="newRevisionId", min_length=1, max_length=64)
    selected_option_id: str = Field(alias="selectedOptionId", min_length=1, max_length=64)
    response_options: list[ResponseOptionV1] | None = Field(
        default=None,
        alias="responseOptions",
    )
    rationale: str = Field(min_length=1, max_length=4096)
    risk_tradeoffs: str = Field(alias="riskTradeoffs", min_length=1, max_length=4096)
    comment: str = Field(default="", max_length=2048)
    modified_by: str = Field(alias="modifiedBy", min_length=1, max_length=128)
    modified_at: UtcTimestamp = Field(alias="modifiedAt")

    @model_validator(mode="after")
    def validate_schema_version(self) -> ProposalModificationV1:
        assert_supported_schema_version("proposal_modification", self.schema_version)
        if self.schema_version != PROPOSAL_MODIFICATION_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported proposal modification schema: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self


class FinalPolicyCheckV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    proposal_id: ProposalId = Field(alias="proposalId")
    proposal_revision_id: str = Field(alias="proposalRevisionId", min_length=1, max_length=64)
    current_revision_id: str = Field(alias="currentRevisionId", min_length=1, max_length=64)
    outcome: PolicyOutcomeV1
    decision: PolicyDecisionV1
    executable: bool
    checked_at: UtcTimestamp = Field(alias="checkedAt")

    @model_validator(mode="after")
    def validate_schema_version(self) -> FinalPolicyCheckV1:
        assert_supported_schema_version("final_policy_check", self.schema_version)
        if self.schema_version != FINAL_POLICY_CHECK_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported final policy check schema: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self


class AuthorizedSimulationCommandV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    proposal_id: ProposalId = Field(alias="proposalId")
    approval_id: ApprovalId = Field(alias="approvalId")
    run_id: RunId = Field(alias="runId")
    scenario_command: ScenarioCommandTemplateV1 = Field(alias="scenarioCommand")
    plugin_id: str = Field(alias="pluginId", min_length=1, max_length=128)
    target_asset_id: str = Field(alias="targetAssetId", min_length=1, max_length=128)
    config: dict[str, Any] = Field(default_factory=dict)
    command_id: str = Field(alias="commandId", min_length=1, max_length=256)
    authorization_token: str = Field(alias="authorizationToken", min_length=1, max_length=256)
    actor_id: str = Field(alias="actorId", min_length=1, max_length=128)

    @model_validator(mode="after")
    def validate_schema_version(self) -> AuthorizedSimulationCommandV1:
        assert_supported_schema_version("authorized_simulation_command", self.schema_version)
        if self.schema_version != AUTHORIZED_SIMULATION_COMMAND_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported authorized simulation command schema: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self


class ExecutionResultV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    executed_action: ExecutedActionV1 = Field(alias="executedAction")
    result_event_id: EventId = Field(alias="resultEventId")
    command_id: str = Field(alias="commandId", min_length=1, max_length=256)
    success: bool
    replayed: bool = False
    message: str = Field(default="", max_length=1024)

    @model_validator(mode="after")
    def validate_schema_version(self) -> ExecutionResultV1:
        assert_supported_schema_version("execution_result", self.schema_version)
        if self.schema_version != EXECUTION_RESULT_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported execution result schema: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self


class ApproveProposalRequestV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    proposal_id: ProposalId = Field(alias="proposalId")
    expected_revision_id: str = Field(alias="expectedRevisionId", min_length=1, max_length=64)
    expected_revision: Revision = Field(alias="expectedRevision")
    comment: str = Field(default="", max_length=2048)
    idempotency_key: str = Field(alias="idempotencyKey", min_length=1, max_length=256)
    actor_id: str | None = Field(default=None, alias="actorId", max_length=128)

    @model_validator(mode="after")
    def validate_schema_version(self) -> ApproveProposalRequestV1:
        assert_supported_schema_version("approve_proposal_request", self.schema_version)
        if self.schema_version != APPROVE_PROPOSAL_REQUEST_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported approve proposal request schema: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self


class RejectProposalRequestV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    proposal_id: ProposalId = Field(alias="proposalId")
    expected_revision_id: str = Field(alias="expectedRevisionId", min_length=1, max_length=64)
    expected_revision: Revision = Field(alias="expectedRevision")
    reason: str = Field(min_length=1, max_length=2048)
    comment: str = Field(default="", max_length=2048)
    idempotency_key: str = Field(alias="idempotencyKey", min_length=1, max_length=256)
    actor_id: str | None = Field(default=None, alias="actorId", max_length=128)

    @model_validator(mode="after")
    def validate_schema_version(self) -> RejectProposalRequestV1:
        assert_supported_schema_version("reject_proposal_request", self.schema_version)
        if self.schema_version != REJECT_PROPOSAL_REQUEST_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported reject proposal request schema: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self


class ModifyProposalRequestV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    proposal_id: ProposalId = Field(alias="proposalId")
    expected_revision_id: str = Field(alias="expectedRevisionId", min_length=1, max_length=64)
    expected_revision: Revision = Field(alias="expectedRevision")
    selected_option_id: str = Field(alias="selectedOptionId", min_length=1, max_length=64)
    response_options: list[ResponseOptionV1] | None = Field(
        default=None,
        alias="responseOptions",
    )
    rationale: str = Field(min_length=1, max_length=4096)
    risk_tradeoffs: str = Field(alias="riskTradeoffs", min_length=1, max_length=4096)
    comment: str = Field(default="", max_length=2048)
    idempotency_key: str = Field(alias="idempotencyKey", min_length=1, max_length=256)
    actor_id: str | None = Field(default=None, alias="actorId", max_length=128)

    @model_validator(mode="after")
    def validate_schema_version(self) -> ModifyProposalRequestV1:
        assert_supported_schema_version("modify_proposal_request", self.schema_version)
        if self.schema_version != MODIFY_PROPOSAL_REQUEST_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported modify proposal request schema: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self


class CancelProposalRequestV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    proposal_id: ProposalId = Field(alias="proposalId")
    expected_revision_id: str = Field(alias="expectedRevisionId", min_length=1, max_length=64)
    expected_revision: Revision = Field(alias="expectedRevision")
    reason: str = Field(min_length=1, max_length=2048)
    idempotency_key: str = Field(alias="idempotencyKey", min_length=1, max_length=256)
    actor_id: str | None = Field(default=None, alias="actorId", max_length=128)

    @model_validator(mode="after")
    def validate_schema_version(self) -> CancelProposalRequestV1:
        assert_supported_schema_version("cancel_proposal_request", self.schema_version)
        if self.schema_version != CANCEL_PROPOSAL_REQUEST_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported cancel proposal request schema: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self


class ApproveProposalResponseV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    approval: ApprovalV1
    final_policy_check: FinalPolicyCheckV1 = Field(alias="finalPolicyCheck")
    execution: ExecutionResultV1 | None = None
    proposal_status: ProposalStatus = Field(alias="proposalStatus")
    replayed: bool = False

    @model_validator(mode="after")
    def validate_schema_version(self) -> ApproveProposalResponseV1:
        assert_supported_schema_version("approve_proposal_response", self.schema_version)
        if self.schema_version != APPROVE_PROPOSAL_RESPONSE_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported approve proposal response schema: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self


class RejectProposalResponseV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    approval: ApprovalV1
    proposal_status: ProposalStatus = Field(alias="proposalStatus")
    replayed: bool = False

    @model_validator(mode="after")
    def validate_schema_version(self) -> RejectProposalResponseV1:
        assert_supported_schema_version("reject_proposal_response", self.schema_version)
        if self.schema_version != REJECT_PROPOSAL_RESPONSE_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported reject proposal response schema: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self


class ModifyProposalResponseV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    modification: ProposalModificationV1
    policy_decision: PolicyDecisionV1 = Field(alias="policyDecision")
    proposal_status: ProposalStatus = Field(alias="proposalStatus")
    replayed: bool = False

    @model_validator(mode="after")
    def validate_schema_version(self) -> ModifyProposalResponseV1:
        assert_supported_schema_version("modify_proposal_response", self.schema_version)
        if self.schema_version != MODIFY_PROPOSAL_RESPONSE_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported modify proposal response schema: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self


__all__ = [
    "ApprovalErrorCode",
    "ApproveProposalRequestV1",
    "ApproveProposalResponseV1",
    "AuthorizedSimulationCommandV1",
    "CancelProposalRequestV1",
    "ExecutionResultV1",
    "FinalPolicyCheckV1",
    "ModifyProposalRequestV1",
    "ModifyProposalResponseV1",
    "ProposalModificationV1",
    "RejectProposalRequestV1",
    "RejectProposalResponseV1",
    "StaleProposalErrorV1",
]

File: packages/contracts-ts/src/approvals.ts
/** Phase 24 approval workflow contracts. */

import { z } from 'zod';

import { approvalSchema, executedActionSchema } from './entities';
import {
  approvalIdSchema,
  eventIdSchema,
  proposalIdSchema,
  revisionSchema,
  runIdSchema,
  utcTimestampSchema,
} from './primitives';
import {
  policyDecisionSchema,
  policyOutcomeSchema,
  responseOptionSchema,
  scenarioCommandTemplateSchema,
} from './proposals';
import {
  APPROVE_PROPOSAL_REQUEST_SCHEMA_VERSION,
  APPROVE_PROPOSAL_RESPONSE_SCHEMA_VERSION,
  AUTHORIZED_SIMULATION_COMMAND_SCHEMA_VERSION,
  CANCEL_PROPOSAL_REQUEST_SCHEMA_VERSION,
  EXECUTION_RESULT_SCHEMA_VERSION,
  FINAL_POLICY_CHECK_SCHEMA_VERSION,
  MODIFY_PROPOSAL_REQUEST_SCHEMA_VERSION,
  MODIFY_PROPOSAL_RESPONSE_SCHEMA_VERSION,
  PROPOSAL_MODIFICATION_SCHEMA_VERSION,
  REJECT_PROPOSAL_REQUEST_SCHEMA_VERSION,
  REJECT_PROPOSAL_RESPONSE_SCHEMA_VERSION,
  STALE_PROPOSAL_ERROR_SCHEMA_VERSION,
} from './versioning';

const schemaVersionCheck = (expected: number) =>
  z
    .number()
    .int()
    .min(1)
    .refine((v) => v === expected);

export const approvalErrorCodeSchema = z.enum([
  'STALE_PROPOSAL',
  'UNAUTHORIZED',
  'POLICY_BLOCKED',
  'VALIDATION_FAILED',
  'CONFLICT',
  'NOT_FOUND',
  'ALREADY_DECIDED',
  'EXECUTION_FAILED',
]);

export const staleProposalErrorSchema = z
  .object({
    schemaVersion: schemaVersionCheck(STALE_PROPOSAL_ERROR_SCHEMA_VERSION),
    code: approvalErrorCodeSchema.default('STALE_PROPOSAL'),
    message: z.string().min(1).max(1024),
    proposalId: proposalIdSchema,
    expectedRevisionId: z.string().min(1).max(64).nullable().optional(),
    actualRevisionId: z.string().min(1).max(64).nullable().optional(),
    expectedRevision: revisionSchema.nullable().optional(),
    actualRevision: revisionSchema.nullable().optional(),
    details: z.record(z.string(), z.unknown()).default({}),
  })
  .strict();

export const proposalModificationSchema = z
  .object({
    schemaVersion: schemaVersionCheck(PROPOSAL_MODIFICATION_SCHEMA_VERSION),
    proposalId: proposalIdSchema,
    previousRevisionId: z.string().min(1).max(64),
    newRevisionId: z.string().min(1).max(64),
    selectedOptionId: z.string().min(1).max(64),
    responseOptions: z.array(responseOptionSchema).nullable().optional(),
    rationale: z.string().min(1).max(4096),
    riskTradeoffs: z.string().min(1).max(4096),
    comment: z.string().max(2048).default(''),
    modifiedBy: z.string().min(1).max(128),
    modifiedAt: utcTimestampSchema,
  })
  .strict();

export const finalPolicyCheckSchema = z
  .object({
    schemaVersion: schemaVersionCheck(FINAL_POLICY_CHECK_SCHEMA_VERSION),
    proposalId: proposalIdSchema,
    proposalRevisionId: z.string().min(1).max(64),
    currentRevisionId: z.string().min(1).max(64),
    outcome: policyOutcomeSchema,
    decision: policyDecisionSchema,
    executable: z.boolean(),
    checkedAt: utcTimestampSchema,
  })
  .strict();

export const authorizedSimulationCommandSchema = z
  .object({
    schemaVersion: schemaVersionCheck(AUTHORIZED_SIMULATION_COMMAND_SCHEMA_VERSION),
    proposalId: proposalIdSchema,
    approvalId: approvalIdSchema,
    runId: runIdSchema,
    scenarioCommand: scenarioCommandTemplateSchema,
    pluginId: z.string().min(1).max(128),
    targetAssetId: z.string().min(1).max(128),
    config: z.record(z.string(), z.unknown()).default({}),
    commandId: z.string().min(1).max(256),
    authorizationToken: z.string().min(1).max(256),
    actorId: z.string().min(1).max(128),
  })
  .strict();

export const executionResultSchema = z
  .object({
    schemaVersion: schemaVersionCheck(EXECUTION_RESULT_SCHEMA_VERSION),
    executedAction: executedActionSchema,
    resultEventId: eventIdSchema,
    commandId: z.string().min(1).max(256),
    success: z.boolean(),
    replayed: z.boolean().default(false),
    message: z.string().max(1024).default(''),
  })
  .strict();

export const approveProposalRequestSchema = z
  .object({
    schemaVersion: schemaVersionCheck(APPROVE_PROPOSAL_REQUEST_SCHEMA_VERSION),
    proposalId: proposalIdSchema,
    expectedRevisionId: z.string().min(1).max(64),
    expectedRevision: revisionSchema,
    comment: z.string().max(2048).default(''),
    idempotencyKey: z.string().min(1).max(256),
    actorId: z.string().max(128).nullable().optional(),
  })
  .strict();

export const rejectProposalRequestSchema = z
  .object({
    schemaVersion: schemaVersionCheck(REJECT_PROPOSAL_REQUEST_SCHEMA_VERSION),
    proposalId: proposalIdSchema,
    expectedRevisionId: z.string().min(1).max(64),
    expectedRevision: revisionSchema,
    reason: z.string().min(1).max(2048),
    comment: z.string().max(2048).default(''),
    idempotencyKey: z.string().min(1).max(256),
    actorId: z.string().max(128).nullable().optional(),
  })
  .strict();

export const modifyProposalRequestSchema = z
  .object({
    schemaVersion: schemaVersionCheck(MODIFY_PROPOSAL_REQUEST_SCHEMA_VERSION),
    proposalId: proposalIdSchema,
    expectedRevisionId: z.string().min(1).max(64),
    expectedRevision: revisionSchema,
    selectedOptionId: z.string().min(1).max(64),
    responseOptions: z.array(responseOptionSchema).nullable().optional(),
    rationale: z.string().min(1).max(4096),
    riskTradeoffs: z.string().min(1).max(4096),
    comment: z.string().max(2048).default(''),
    idempotencyKey: z.string().min(1).max(256),
    actorId: z.string().max(128).nullable().optional(),
  })
  .strict();

export const cancelProposalRequestSchema = z
  .object({
    schemaVersion: schemaVersionCheck(CANCEL_PROPOSAL_REQUEST_SCHEMA_VERSION),
    proposalId: proposalIdSchema,
    expectedRevisionId: z.string().min(1).max(64),
    expectedRevision: revisionSchema,
    reason: z.string().min(1).max(2048),
    idempotencyKey: z.string().min(1).max(256),
    actorId: z.string().max(128).nullable().optional(),
  })
  .strict();

export const approveProposalResponseSchema = z
  .object({
    schemaVersion: schemaVersionCheck(APPROVE_PROPOSAL_RESPONSE_SCHEMA_VERSION),
    approval: approvalSchema,
    finalPolicyCheck: finalPolicyCheckSchema,
    execution: executionResultSchema.nullable().optional(),
    proposalStatus: z.enum(['pending', 'approved', 'rejected', 'executed', 'cancelled']),
    replayed: z.boolean().default(false),
  })
  .strict();

export const rejectProposalResponseSchema = z
  .object({
    schemaVersion: schemaVersionCheck(REJECT_PROPOSAL_RESPONSE_SCHEMA_VERSION),
    approval: approvalSchema,
    proposalStatus: z.enum(['pending', 'approved', 'rejected', 'executed', 'cancelled']),
    replayed: z.boolean().default(false),
  })
  .strict();

export const modifyProposalResponseSchema = z
  .object({
    schemaVersion: schemaVersionCheck(MODIFY_PROPOSAL_RESPONSE_SCHEMA_VERSION),
    modification: proposalModificationSchema,
    policyDecision: policyDecisionSchema,
    proposalStatus: z.enum(['pending', 'approved', 'rejected', 'executed', 'cancelled']),
    replayed: z.boolean().default(false),
  })
  .strict();

export type ApprovalErrorCode = z.infer<typeof approvalErrorCodeSchema>;
export type StaleProposalErrorV1 = z.infer<typeof staleProposalErrorSchema>;
export type ProposalModificationV1 = z.infer<typeof proposalModificationSchema>;
export type FinalPolicyCheckV1 = z.infer<typeof finalPolicyCheckSchema>;
export type AuthorizedSimulationCommandV1 = z.infer<typeof authorizedSimulationCommandSchema>;
export type ExecutionResultV1 = z.infer<typeof executionResultSchema>;
export type ApproveProposalRequestV1 = z.infer<typeof approveProposalRequestSchema>;
export type RejectProposalRequestV1 = z.infer<typeof rejectProposalRequestSchema>;
export type ModifyProposalRequestV1 = z.infer<typeof modifyProposalRequestSchema>;
export type CancelProposalRequestV1 = z.infer<typeof cancelProposalRequestSchema>;
export type ApproveProposalResponseV1 = z.infer<typeof approveProposalResponseSchema>;
export type RejectProposalResponseV1 = z.infer<typeof rejectProposalResponseSchema>;
export type ModifyProposalResponseV1 = z.infer<typeof modifyProposalResponseSchema>;
